### PR TITLE
Fix NPE in TypeHashVisitor when use null in union-type 

### LINKED ISSUE #34556: NPE in `TypeHashVisitor` when use `null` in union-type
**Description:**
NPE occurs in `TypeHashVisitor` when the compiled source contains a function type with a parameter of union-type that defined as follows
```ballerina
public type TestType string|int?[]|null;

public function doThis(TestType res) {
    
}
```

**Steps to reproduce:**
1. Include the above code into the ballerina project. Build it and push it to the `local` repository
2. Create a ballerina file and try using the `doThis` function by passing it to a function call. (see the example below)
```ballerina
import org/testType;

public function main() {
    func1(testType:doThis);
}

function func1(function (testType:TestType resp) callback) {
}
```
3. Run 

**Affected Versions:**
`Ballerina 2201.0.0 (Swan Lake) Snapshot build 1`, `stage-swan-lake` branch

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by rpjayasekara**
Fixes #34556

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_import_type_with_null/test_import_project/test_type.bal [hunk 1]
```diff
@@ -0,0 +1,5 @@
+public type TestType string|null;
+
+public function doThis(TestType res) {
+
+}
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_import_type_with_null/test_import_project/test_type.bal:5]: Add new line here

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.Arrays;
+
+/**
+ * Test cases for import user defined type.
+ */
+public class ImportUserDefinedTypeTest {
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java:15]: Shall we add some json related tests also, including nil values in them

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package org.ballerinalang.test.bala.types;
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java:1]: Add copyright headers

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_import_type_with_null/main.bal [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import testorg/testType;
+
+function dummyTest() {
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_import_type_with_null/main.bal:3]: Check and add some bala tests with positive/negative type assignments for nil literal if not already there

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java [hunk 5]
```diff
@@ -88,7 +88,7 @@ public NodeKind getKind() {
     @Override
     public String toString() {
         // To distinguish between `()` and `null`
-        return value == null ? originalValue : String.valueOf(value);
+        return value == null ? (originalValue == null ? "null" : originalValue) : String.valueOf(value);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java:91]: Is this the correct fix? Isn't the original check intended to cover exactly this scenario? Do we know why `value` is not null but `originalValue` is null?

I believe this is related to https://github.com/ballerina-platform/ballerina-lang/issues/34588.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java (hunk 2 above)
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java:15]: I don't think this name properly reflects what we are testing. There are other tests that (or rather pretty much all tests) import modules and use types defined in them.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import org.ballerinalang.test.CompileResult;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.Arrays;
+
+/**
+ * Test cases for import user defined type.
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java:13]: Let's also add the `@since` annotation.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ */
+public class ImportUserDefinedTypeTest {
+    private CompileResult result;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/ImportUserDefinedTypeTest.java:16]: Given that `result` is used in only one test method we can make it a private variable there and get rid of the `tearDown` method,

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_import_type_with_null/test_import_project/Ballerina.toml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+[package]
+org = "testorg"
+name = "testType"
+version = "0.1.0"
+
+[build-options]
+observabilityIncluded = true
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_import_type_with_null/test_import_project/Ballerina.toml:7]: Not required.

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_import_type_with_null/test_import_project/test_type.bal [hunk 9]
```diff
@@ -0,0 +1,5 @@
+public type TestType string|null;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_import_type_with_null/test_import_project/test_type.bal:1]: We use underscores for file names and dashes for directories.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java (hunk 5 above)
Comment (rpjayasekara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java:91]: Both `originalValue` and `Value` gets `null` when reading, since ` BIRTypeWriter` does not write when the type tag is `NIL`

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java:91]: `null` or `"null"`?

Comment (rpjayasekara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java:91]: `null`

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java:91]: My original comment should have been

> Do we know why value is `null` but `originalValue` is not `"null"`?

Ideally, the behaviour should be the same in normal compilation and via BIR symbol enter.

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java:91]: Actually I think we have an issue differentiating between `"null"` and `null` anyway, which is probably also the reason for https://github.com/ballerina-platform/ballerina-lang/issues/34588.

With this fix, wouldn't we see this issue further?

Comment (rpjayasekara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java:91]: Current implementation neglect writing value when the type is `NIL`
https://github.com/ballerina-platform/ballerina-lang/blob/fdb8d28e9454125877cb5405043dd92c7faf754b/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java#L562

Comment (rpjayasekara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java:91]: Yeah IMO differentiating "null" and null is different issue compared to this

> Actually I think we have an issue differentiating between `"null"` and `null` anyway, which is probably also the reason for #34588.
> 
> With this fix, wouldn't we see this issue further?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangLiteral.java:91]: > Current implementation neglect writing value when the type is `NIL`
> 
> https://github.com/ballerina-platform/ballerina-lang/blob/fdb8d28e9454125877cb5405043dd92c7faf754b/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java#L562

Even then we can create the value properly (as what was there when generating the BIR) at BIR symbol enter, right?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
+/*
+ *   Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:2]: This has to be 2022

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+import org.ballerinalang.test.CompileResult;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.Arrays;
+
+/**
+ * Test cases for import user defined type.
+ * @since 1.3.0
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:30]: @since 2.0.0?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/json/JSONTest.java [hunk 12]
```diff
@@ -118,6 +118,13 @@ public void testParseNull() {
         Assert.assertNull(returns[0]);
     }
 
+    @Test(description = "Get JSON null from a string")
+    public void testJSONWithNullValues() {
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/json/JSONTest.java:122]: Shall we move this assertion to bal file itself?

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/negative_assignment.bal [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+
+function testNegativeAssignment() returns string{
+    string a = func1(testType:doThis);
+}
+
+function func1(function (testType:TestType resp) returns testType:TestType callback) returns testType:TestType{
+    testType:TestType a = null;
+    testType:TestType b = callback(a);
+    return b;
+}
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/negative_assignment.bal:11]: newline

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/negative_assignment.bal [hunk 14]
```diff
@@ -0,0 +1,11 @@
+import testorg/testType;
+
+function testNegativeAssignment() returns string{
+    string a = func1(testType:doThis);
+}
+
+function func1(function (testType:TestType resp) returns testType:TestType callback) returns testType:TestType{
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/negative_assignment.bal:7]:
```suggestion
function func1(function (testType:TestType resp) returns testType:TestType callback) returns testType:TestType {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/negative_assignment.bal (hunk 13 above)
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/negative_assignment.bal:11]:
```suggestion
}

```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/test_type.bal [hunk 15]
```diff
@@ -0,0 +1,5 @@
+public type TestType string|null;
+
+public function doThis(TestType res) returns TestType{
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/test_type.bal:3]:
```suggestion
public function doThis(TestType res) returns TestType {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/test_type.bal [hunk 16]
```diff
@@ -0,0 +1,5 @@
+public type TestType string|null;
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/test_type.bal:1]: Shall we add one more tests, like importing `TestType` from a different module?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java (hunk 10 above)
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:2]:
```suggestion
 *   Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java (hunk 11 above)
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:30]:
```suggestion
 * @since 2.0.0
```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal [hunk 17]
```diff
@@ -0,0 +1,11 @@
+import testorg/testType;
+
+function testPositiveAssignment(){
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal:3]:
```suggestion
function testPositiveAssignment() {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+    testType:TestType a = func1(testType:doThis);
+}
+
+function func1(function (testType:TestType resp) returns testType:TestType callback) returns testType:TestType{
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal:7]:
```suggestion
function func1(function (testType:TestType resp) returns testType:TestType callback) returns testType:TestType {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/jsontype/json-test.bal [hunk 19]
```diff
@@ -30,6 +30,11 @@ function testJSONArrayToJsonAssignment() returns (json) {
     return j2;
 }
 
+function testJSONWithNullValues() returns string {
+    json j = {"name":{"fname":"Jack", "lname":"Taylor"}, "state": null, "age":()};
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/jsontype/json-test.bal:34]:
```suggestion
    json j = {"name": {"fname": "Jack", "lname": "Taylor"}, "state": null, "age": ()};
```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/types.bal [hunk 20]
```diff
@@ -0,0 +1 @@
+public type A int?[]|null;
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/types.bal:1]:
```suggestion
public type A int?[]|null;

```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/test_type.bal (hunk 16 above)
Comment (gimantha) [reply tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/test_type.bal:1]: @rpjayasekara sorry, ignore this comment. It is already in a different module right?

Comment (rpjayasekara) [reply tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/test_type.bal:1]: Yeah

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.Arrays;
+
+/**
+ * Test cases for import user defined type.
+ * @since 2.0.0
+ */
+public class UnionTypeWithNullLiteralTest {
+
+    private CompileResult result;
+    private CompileResult negativeCompileResult;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:35]: Can't these be local variables within the test methods?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Test
+    public void unionTypeWithNullLiteralValueAssignment() {
+        Assert.assertEquals(result.getErrorCount(), 0, Arrays.asList(result.getDiagnostics()).toString());
+    }
+
+    @Test
+    public void unionTypeWithNullLiteralValueAssignmentNegative() {
+        Assert.assertEquals(negativeCompileResult.getErrorCount(), 1,
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:53]: We have to assert the error.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/json/JSONTest.java [hunk 23]
```diff
@@ -118,6 +118,11 @@ public void testParseNull() {
         Assert.assertNull(returns[0]);
     }
 
+    @Test(description = "Get JSON with null values")
+    public void testJSONWithNullValues() {
+        BValue[] returns = BRunUtil.invoke(compileResult, "testJSONWithNullValues");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/json/JSONTest.java:123]:
```suggestion
        BRunUtil.invoke(compileResult, "testJSONWithNullValues");
```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal [hunk 24]
```diff
@@ -0,0 +1,15 @@
+import testorg/testType;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal:1]: Add licence headers.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+import org.ballerinalang.test.BCompileUtil;
+import org.ballerinalang.test.CompileResult;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.Arrays;
+
+/**
+ * Test cases for import user defined type.
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:29]:
```suggestion
 * Test cases for importing a user-defined union type containing the null literal.
 * 
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+ * @since 2.0.0
+ */
+public class UnionTypeWithNullLiteralTest {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:32]:
```suggestion
public class UnionTypeWithNullLiteralBalaTest {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/jsontype/json-test.bal [hunk 27]
```diff
@@ -47,6 +52,13 @@ public function testAssignabilityToUnion() returns (MyJson | xml) {
     return c;
 }
 
+function assertEquality(any actual, any expected) {
+    if actual is anydata && expected is anydata && actual == expected {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/jsontype/json-test.bal:56]:
```suggestion
function assertEquality(anydata actual, anydata expected) {
    if actual == expected {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+    public void setup() {
+        BCompileUtil.compileAndCacheBala("test-src/bala/test_projects/test-union-type-with-null-literal/type-export");
+
+        this.result = BCompileUtil.compile("test-src/bala/test_projects/test-union-type-with-null-literal/main.bal");
+        this.negativeCompileResult = BCompileUtil.compile(
+                "test-src/bala/test_projects/test-union-type-with-null-literal/negative_assignment.bal");
+    }
+
+    @Test
+    public void unionTypeWithNullLiteralValueAssignment() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:47]:
```suggestion
    public void testUnionTypeWithNullLiteral() {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+        Assert.assertEquals(result.getErrorCount(), 0, Arrays.asList(result.getDiagnostics()).toString());
+    }
+
+    @Test
+    public void unionTypeWithNullLiteralValueAssignmentNegative() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:52]:
```suggestion
    public void testUnionTypeWithNullLiteralNegative() {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/json/JSONTest.java [hunk 30]
```diff
@@ -118,6 +118,11 @@ public void testParseNull() {
         Assert.assertNull(returns[0]);
     }
 
+    @Test(description = "Get JSON with null values")
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/json/JSONTest.java:121]: How is this test relevant to this fix? This works already also.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+    private CompileResult result;
+    private CompileResult negativeCompileResult;
+
+    @BeforeClass
+    public void setup() {
+        BCompileUtil.compileAndCacheBala("test-src/bala/test_projects/test-union-type-with-null-literal/type-export");
+
+        this.result = BCompileUtil.compile("test-src/bala/test_projects/test-union-type-with-null-literal/main.bal");
+        this.negativeCompileResult = BCompileUtil.compile(
+                "test-src/bala/test_projects/test-union-type-with-null-literal/negative_assignment.bal");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:43]: The test files usually go in a separate folder. Not in `test-src/bala`.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:43]: We wouldn't need `type-export` separately then. Please check what's done with other bala tests.

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/types.bal (hunk 20 above)
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/type-export/types.bal:1]:
```suggestion
public type TestType2 int?[]|null;
```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal [hunk 32]
```diff
@@ -0,0 +1,15 @@
...
+
+function testPositiveAssignment() {
+    testType:TestType a = func1(testType:doThis);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal:4]:
```suggestion
    testType:TestType _ = func1(testType:doThis);
```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal [hunk 33]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+function func1(function (testType:TestType resp) returns testType:TestType callback) returns testType:TestType {
+    testType:TestType a = null;
+    testType:TestType b = callback(a);
+    return b;
+}
+
+function testImportTypeFromSeparateModule() {
+    testType:A a = null;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal:14]:
```suggestion
    testType:A _ = null;
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java [hunk 34]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Test
+    public void unionTypeWithNullLiteralValueAssignment() {
+        Assert.assertEquals(result.getErrorCount(), 0, Arrays.asList(result.getDiagnostics()).toString());
+    }
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralTest.java:49]: In the positive test, shouldn't we do some assertion?

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal [hunk 35]
```diff
@@ -0,0 +1,15 @@
...
+}
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal:15]: In the positive test, shouldn't we do some assertion?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/json/JSONTest.java (hunk 30 above)
Comment (rpjayasekara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/json/JSONTest.java:121]: This test was added since to check weather `null` and `()` is working correctly in JSON context after changing the `toString` method. Since now that changed is not included in this PR anymore, I will remove this

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal (hunk 35 above)
Comment (rpjayasekara) [reply tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal:15]: I did not any assertions because problem was triggered when importing the type. Since now the type can import and can use without any errors, I though of not to perform assertion to valideate

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test-union-type-with-null-literal/main.bal:15]: IMO, it is still better to assert the value, since the crash meant we weren't testing the value part either.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralBalaTest.java [hunk 36]
```diff
@@ -0,0 +1,57 @@
...
+import org.ballerinalang.test.BRunUtil;
+import org.ballerinalang.test.CompileResult;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.ballerinalang.test.BAssertUtil.validateError;
+
+/**
+ * Test cases for importing a user-defined union type containing the null literal.
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/UnionTypeWithNullLiteralBalaTest.java:30]:
```suggestion
 * Test cases for importing a user-defined union type containing the null literal.
 * 
```

**Review by MaryamZi — APPROVED** [review]
